Avoid concurrent shutdown of Infinispan caches to avoid lockup.

ahus1 opened the pull request:
Closes #10016

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

ahus1 committed:
Workaround for deadlock when shutting down Infinispan in 12.1.7.Final.

This is tracked in upstream issue https://issues.redhat.com/browse/ISPN-13664

Closes #10016

ahus1 committed:
Ensure that Infinispan shutdowns correctly at the end of the tests. Report any exceptions within another thread as a test failure.

Adding additional information like a thread dump when it doesn't shutdown as expected.

Closes #10016

hmlnarik commented on testsuite/model/src/test/java/org/keycloak/testsuite/model/KeycloakModelTest.java:
```diff
@@ -366,7 +383,18 @@ public static void inIndependentFactories(int numThreads, int timeoutSeconds, Ru
+                    for (StackTraceElement traceElement : thread.getStackTrace()) {
+                        sb.append("\tat ").append(traceElement).append("\n");
+                    }
+                    return sb.toString();
```
Can be replaced with a less talkative stack trace, see `StackUtil` methods

ahus1 replied:
@hmlnarik - I had a look at `StackUtil` and it only reads the current thread's stack trace. Here I'm interested in other threads stack traces. Therefore I'd like to keep the code. In the most recent version I'm adding more information about other threads running on the JVM.

ahus1 replied:
The code evolved from this even further. As the code it relates to is obsolete, I resolve this.

ahus1 commented on testsuite/model/src/test/java/org/keycloak/testsuite/model/session/OfflineSessionPersistenceTest.java:
```diff
@@ -130,6 +131,7 @@ public void testPersistenceSingleNode() {
     }
 
     @Test(timeout = 90 * 1000)
+    @Ignore("test currently fails as offline sessions are not loaded by default any more?!")
```
@martin-kanis - I've been updating inIndependentFactories() so that any assert wrapped inside it will make the test fail. Due to this change, the following code now fails when "keycloak.model.parameters=Infinispan,Jpa" are set.

I now wonder: is this due to offline sessions no longer being loaded by default, one of the PR's that went through your hands lately?

```
inIndependentFactories(4, 30, () -> assertOfflineSessionsExist(realmId, clientSessionIds.keySet()));
```

martin-kanis replied:
After quick debugging it seems no offline user sessions are created. I suspect there's a bug in this block https://github.com/keycloak/keycloak/pull/10288/files#diff-ab99606564bc3505e0948722bd230a49e89252a990a279c078445d32c4d259c8R150-R161
When I evaluate `session.sessions().getOfflineUserSessionsStream(realm, session.users().getUserById(realm, userIds.get(0))).collect(Collectors.toList())` at the end of the block, there are no sessions.

ahus1 replied:
I looked at the test again today, and found that the stream returned by createOfflineSessions was never consumed, therefore no offline sessions were created. Will push a change once the current run completes (don't want to interrupt it).

ahus1 commented on testsuite/model/src/test/java/org/keycloak/testsuite/model/session/OfflineSessionPersistenceTest.java:
```diff
@@ -282,6 +286,8 @@ private String createOfflineClientSession(String offlineUserSessionId, String cl
     }
 
     @Test(timeout = 90 * 1000)
+    @RequireProvider(UserSessionPersisterProvider.class)
+    @RequireProvider(value = UserSessionProvider.class, only = InfinispanUserSessionProviderFactory.PROVIDER_ID)
```
@martin-kanis - same as above. Your help is appreciated...

ahus1 commented on testsuite/model/src/test/java/org/keycloak/testsuite/model/infinispan/CacheExpirationTest.java:
```diff
@@ -71,12 +74,16 @@ public void testCacheExpiration() throws Exception {
         // Ensure that instance counting works as expected, there should be at least two instances in memory now.
         // Infinispan server is decoding the client request before processing the request at the cache level,
         // therefore there are sometimes three instances of AuthenticationSessionAuthNoteUpdateEvent class in the memory
```
@martin-kanis - I now see that when running on GitHub actions, when there are three instances in this location, there will be at the end one instance that will not be garbage collected. 

Looking at the git history, that these two lines were added by you. I wonder if you might recall where this one instance is kept, and the circumstances this occurs in. Even if you don't, it would be great if you could comment if it is ok for the test to succeed if this one instance is still on the heap.

martin-kanis replied:
I discussed this problem with Ryan from Infinispan team. His exact explanation was:
"This is just a transient use of the class when the server is decoding the client request before processing the request at the cache level, hence why you only see it sometimes. I don't think this is anything to be concerned about."

I have a screenshot from heap dump which showing the references of this instance, if you are interested. 

So I don't think we need to worry about it in the test.

ahus1 replied:
Alright, I'll then re-write the test pass with one instance still in memory in the case when a third instance appeared after inserting the first two instances. When running the test on GitHub Actions, that one instance didn't disappear, it seems to be bound to some context so it can't be garbage collected. 

Please briefly confirm that you think this is the right way to go, and I'll then resolve this conversation.

martin-kanis replied:
Sounds good

martin-kanis commented on testsuite/model/src/test/java/org/keycloak/testsuite/model/session/OfflineSessionPersistenceTest.java:
```diff
@@ -164,6 +166,8 @@ public void testPersistenceMultipleNodesClientSessionAtSameNode() throws Interru
     }
 
     @Test(timeout = 90 * 1000)
+    @RequireProvider(UserSessionPersisterProvider.class)
+    @RequireProvider(value = UserSessionProvider.class, only = InfinispanUserSessionProviderFactory.PROVIDER_ID)
```
This is weird one. There's something wrong with map-[realms|user-sessions|client-sessions]-mapStorage-f:ROOT:concurrenthashmap.json file where corresponding storages are serialized. There's probably something missing and I guess it has something to do with concurrency. 
Problem can be reproduced by running `mvn test -f testsuite/model/pom.xml -Phot-rod -Dkeycloak.logging.level=debug -Dtest=OfflineSessionPersistenceTest#testPersistenceMultipleNodesClientSessionsAtRandomNode -Dmaven.surefire.debug` with breakpoint at `ConcurrentHashMapStorageProviderFactory.java:234`

```
Caused by: java.lang.RuntimeException: com.fasterxml.jackson.databind.JsonMappingException: Unexpected end-of-input within/between Object entries
 at [Source: (File); line: 12846, column: 26] (through reference chain: java.util.LinkedList[637])
	at org.keycloak.models.map.storage.chm.ConcurrentHashMapStorageProviderFactory.loadMap(ConcurrentHashMapStorageProviderFactory.java:234)
	at org.keycloak.models.map.storage.chm.ConcurrentHashMapStorageProviderFactory.lambda$getStorage$1(ConcurrentHashMapStorageProviderFactory.java:263)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1705)
	at org.keycloak.models.map.storage.chm.ConcurrentHashMapStorageProviderFactory.getStorage(ConcurrentHashMapStorageProviderFactory.java:263)
	at org.keycloak.models.map.storage.chm.ConcurrentHashMapStorageProvider.getStorage(ConcurrentHashMapStorageProvider.java:43)
	at org.keycloak.models.map.userSession.MapUserSessionProviderFactory.create(MapUserSessionProviderFactory.java:86)
	at org.keycloak.models.map.userSession.MapUserSessionProviderFactory.create(MapUserSessionProviderFactory.java:43)
	at org.keycloak.services.DefaultKeycloakSession.getProvider(DefaultKeycloakSession.java:316)
	at org.keycloak.services.DefaultKeycloakSession.sessions(DefaultKeycloakSession.java:455)
	at org.keycloak.testsuite.model.session.OfflineSessionPersistenceTest.lambda$createOfflineClientSession$28(OfflineSessionPersistenceTest.java:282)
	at org.keycloak.testsuite.model.KeycloakModelTest.lambda$withRealm$13(KeycloakModelTest.java:598)
	at org.keycloak.testsuite.model.KeycloakModelTest.lambda$inComittedTransaction$11(KeycloakModelTest.java:563)
	at org.keycloak.testsuite.model.KeycloakModelTest.lambda$inComittedTransaction$12(KeycloakModelTest.java:580)
	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:239)
	at org.keycloak.testsuite.model.KeycloakModelTest.inComittedTransaction(KeycloakModelTest.java:568)
	at org.keycloak.testsuite.model.KeycloakModelTest.inComittedTransaction(KeycloakModelTest.java:563)
	at org.keycloak.testsuite.model.KeycloakModelTest.withRealm(KeycloakModelTest.java:595)
	at org.keycloak.testsuite.model.session.OfflineSessionPersistenceTest.createOfflineClientSession(OfflineSessionPersistenceTest.java:281)
	at org.keycloak.testsuite.model.session.OfflineSessionPersistenceTest.lambda$testPersistenceMultipleNodesClientSessionsAtRandomNode$19(OfflineSessionPersistenceTest.java:192)
	at org.keycloak.testsuite.model.KeycloakModelTest.lambda$inIndependentFactories$4(KeycloakModelTest.java:373)
	at org.keycloak.testsuite.model.KeycloakModelTest.inIndependentFactory(KeycloakModelTest.java:472)
```

ahus1 replied:
Could it be that each simulated Keycloak instance, the CHM stores should point at is own set of files, but at the moment in this case of concurrent tests they all try to read and write the same file? 

Should that be fixed, so that each of them points at their own set of files, or should the test be ignored for the CHM stores?

martin-kanis replied:
That might be the case but I'm not sure why we see it only in `testPersistenceMultipleNodesClientSessionAtSameNode` and `testPersistenceMultipleNodesClientSessionsAtRandomNode` tests and not in other tests that use `inIndependentFactories`. I'm also curious about @hmlnarik's view on this.

hmlnarik replied:
> Could it be that each simulated Keycloak instance, the CHM stores should point at is own set of files, but at the moment in this case of concurrent tests they all try to read and write the same file?

I believe this is the case. Would this vanish with synchronizing the `loadMap` / `storeMap` methods on `ConcurrentHashMapStorageProviderFactory.class`?

ahus1 replied:
@hmlnarik - yes, the error messages I see above will then vanish. Then the test fails as the number of stored sessions deviates from the actual number of sessions. IMHO testPersistenceMultipleNodesClientSessionsAtRandomNode and testPersistenceMultipleNodesClientSessionAtSameNode would deliver sensible results once they run against a clustered store (either HotRot or Infinispan). WDYT?

hmlnarik replied:
Thank you, then it confirms that it is indeed the cause.

Because of this, I agree with disabling the test for CHM map storage

ahus1 replied:
... looking a bit more closely, HotRod doesn't provide offline session support yet, so Infinispan is the only truly supported multi-node store?

ahus1 replied:
Didn't see your previous comment when writing my comment. Thanks @hmlnarik - resolving this.

ahus1 commented:
Modeltest was in the green! Close/Open to trigger pipeline.

ahus1 commented:
This change now combines multiple optimizations in logging and error handling for the model tests related to Infinispan. The tests work now more reliably, at the same time there might be more problematic things in the code. 

A review feedback would include which changes to keep in this PR and which not, and maybe to split it into multiple PRs.

martin-kanis commented on testsuite/model/src/test/java/org/keycloak/testsuite/model/infinispan/CacheExpirationTest.java:
```diff
@@ -54,6 +53,9 @@
     public void testCacheExpiration() throws Exception {
 
+        log.debug("Check if previous instances of the class exist on the heap");
+        getNumberOfInstancesOfClass(AuthenticationSessionAuthNoteUpdateEvent.class);
```
@ahus1 This is here just for displaying number of instances in memory or does it have any additional purpose?

ahus1 replied:
@martin-kanis - yes, this is the intent. No other purpose. Please comment how I should change this code to make this more clear.

martin-kanis replied:
I see. Maybe put the `getNumberOfInstancesOfClass` directly into log.debug like `log.debugf("Number of previous instances of the class on the heap:  %d", getNumberOfInstancesOfClass(AuthenticationSessionAuthNoteUpdateEvent.class));`?

martin-kanis replied:
done, thank you!

martin-kanis commented:
@ahus1 Thanks for the improvements. It looks good to me.
As Hynek proposed on the call, it would be good to extract change related to `PersistenceManagerImpl.java` to a separate commit. 
Another thing to consider is to introduce this method 
```
protected void awaitLatch(CountDownLatch latch) {
        try {
            latch.await();
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    }
```
into the KeycloakModelTest as a util method and use it across model tests. It became a common pattern in the model test. I introduced the method in `UserSessionProviderOfflineModelTest` but it's private there. The refactoring doesn't need to be included in this PR as it's not strictly related to this issue.

ahus1 commented:
@martin-kanis - thanks for pointing me to the existing latch logic; I missed it as I was looking only at test failures. 

I took this one step further than you suggested and made the start/stop latch the default for all inIndependentFactories() as it looked strange to me to do it for some but not all of them. The awaitLatch() stayed where it was as it is still used in one place. I also ensured that the interrupted flag for the thread is set properly: otherwise the next await or sleep might not see it.

I pushed the changes, the build is still running. Please re-review.

ahus1 commented:
The test on GHA failed due to a transient (?) problem when communicating with LDAP. 

``` 
[ERROR] org.keycloak.testsuite.model.UserModelTest.testAddDirtyRemoveFederationUsersInTheSameGroupConcurrent  Time elapsed: 2.35 s  <<< ERROR!
org.keycloak.models.ModelException: Could not unbind DN [uid=user-68,ou=People,dc=keycloak,dc=org]
   at org.keycloak.storage.ldap.idm.store.ldap.LDAPOperationManager.destroySubcontext(LDAPOperationManager.java:475)
   at org.keycloak.storage.ldap.idm.store.ldap.LDAPOperationManager.access$100(LDAPOperationManager.java:67)
   at org.keycloak.storage.ldap.idm.store.ldap.LDAPOperationManager$1.execute(LDAPOperationManager.java:160)
   at org.keycloak.storage.ldap.idm.store.ldap.LDAPOperationManager$1.execute(LDAPOperationManager.java:153)
   at org.keycloak.storage.ldap.idm.store.ldap.LDAPOperationManager.execute(LDAPOperationManager.java:731)
   at org.keycloak.storage.ldap.idm.store.ldap.LDAPOperationManager.execute(LDAPOperationManager.java:711)
   at org.keycloak.storage.ldap.idm.store.ldap.LDAPOperationManager.execute(LDAPOperationManager.java:706)
   at org.keycloak.storage.ldap.idm.store.ldap.LDAPOperationManager.removeEntry(LDAPOperationManager.java:153)
   at org.keycloak.storage.ldap.idm.store.ldap.LDAPIdentityStore.remove(LDAPIdentityStore.java:234)
   at org.keycloak.storage.ldap.LDAPStorageProvider.removeUser(LDAPStorageProvider.java:322)
   at org.keycloak.testsuite.model.UserModelTest.lambda$testAddDirtyRemoveFederationUsersInTheSameGroupConcurrent$15(UserModelTest.java:216)
   at java.base/java.util.stream.Streams$RangeIntSpliterator.forEachRemaining(Streams.java:104)
   at java.base/java.util.stream.IntPipeline$Head.forEach(IntPipeline.java:593)
   at org.keycloak.testsuite.model.UserModelTest.lambda$testAddDirtyRemoveFederationUsersInTheSameGroupConcurrent$16(UserModelTest.java:214)
   at org.keycloak.testsuite.model.KeycloakModelTest.lambda$withRealm$13(KeycloakModelTest.java:631)
   at org.keycloak.testsuite.model.KeycloakModelTest.lambda$inComittedTransaction$11(KeycloakModelTest.java:596)
   at org.keycloak.testsuite.model.KeycloakModelTest.lambda$inComittedTransaction$12(KeycloakModelTest.java:613)
   at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:239)
   at org.keycloak.testsuite.model.KeycloakModelTest.inComittedTransaction(KeycloakModelTest.java:601)
   at org.keycloak.testsuite.model.KeycloakModelTest.inComittedTransaction(KeycloakModelTest.java:596)
   at org.keycloak.testsuite.model.KeycloakModelTest.withRealm(KeycloakModelTest.java:628)
   at org.keycloak.testsuite.model.UserModelTest.testAddDirtyRemoveFederationUsersInTheSameGroupConcurrent(UserModelTest.java:211)
   at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
   at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
``` 

Added a retry, also stabilized some more tests and fixed some of the logging so it doesn't log to STDERR and pollutes the GHA: 572c769b1e11977f99deeb73338ab31e8700a3ec

Let's see what the next run brings, and see the results tomorrow.

ahus1 commented:
@martin-kanis - Model tests are green, ready to continue your review.

martin-kanis commented:
@ahus1 Thank you for the update. I see the logs from model tests and they are very big. Around 107 MB so I need to download them because my browser can't handle it. There are many stacktraces related to Infinispan. So I am wondering if we can get rid of them. WDYT?

ahus1 commented:
@martin-kanis - reducing the amount of logs gets a +1 from me. I'd like to have it in a separate ticket, and I'm happy to look into it after this issue. Please open an issue an link it to this one.

martin-kanis commented:
> @martin-kanis - reducing the amount of logs gets a +1 from me. I'd like to have it in a separate ticket, and I'm happy to look into it after this issue. Please open an issue an link it to this one.

Here it is https://github.com/keycloak/keycloak/issues/10588. Thank you!

martin-kanis reviewed (approved):
Thanks @ahus1. Looks good.

hmlnarik reviewed (approved):
Thanks for the PR!
